uri-scheme.md

```diff
@@ -0,0 +1,35 @@
+# URI Scheme
+
+The `wormhole` URI Scheme is used to encode a wormhole code as a URI. This can then be used to generate QR codes, or be opened by the platform URI handler to open a supporting client.
+
+The general format looks like this, and assumes default values for all query fields:
+`wormhole:{code}`
+
+It can be extended by appending a [query string](https://en.wikipedia.org/wiki/Query_string).
+
+## Query fields
+Applications MUST parse all query fields specified below and fail if they contain unknown fields or unsupported parameter values. Query values are [URL encoded](https://en.wikipedia.org/wiki/Percent-encoding).
+
+* `version`: The version of the URI scheme.  No other value than `0` is accepted.
+Default: 0
+
+* `rendezvous`: The rendezvous server and protocol to use, including its port.
+Default: `ws://relay.magic-wormhole.io:4000/v1`
+
+* `type`: The type of operation requested. Valid values are:
```

Comment:
What would be the use case for this? Normally the sender generates the code and the receiver enters it, why should it be any different when encoding it in a URI?

Reply:
Not all devices have a camera. It might be convenient to generate the code on a computer for example, scan it from a phone, select the file and send it right away.

Reply:
Interesting. I'm not opposed, but I'd like to discuss this a bit more since it is the only feature that I wouldn't consider as trivial.

To reiterate: The URL is generated by the sender and the receiver scans it. This requires the receiver to have a camera. Otherwise, the code must be transferred as text, either by typing it in old style or by sending over the link (you could also send the code without link, either way). Your proposal here is that when the receiver does not have a camera (e.g. phone -> desktop transfer), the code is generated on receiver side and the sender then can scan it.

Protocol wise this should be fine and easy enough to implement with most libraries. But how do you imagine the UI side, both for CLI and GUI?

Reply:
For my application I would probably have a QR button next to the code entry on the receive tab which would generate a code that could be scanned by the other side. After scanning the code the sending application would understand that this is actually a request and display a message with a button to open up a file picker. After picking a file the file would be sent automatically.

I attached a few mockup screenshots, I would probably change a few things up, but you get the idea.

It is definitely more work to implement for everyone, and might also create a small amount of confusion. But in my opinion this would make the very common workflow of "I have something on my phone and I need it on my computer" way easier.

Requesting side:
![image](https://user-images.githubusercontent.com/121945/168468890-efc1966e-0150-486f-b380-35c92b2ae5e9.png)
![image](https://user-images.githubusercontent.com/121945/168469214-7181d14e-8f25-49a2-8ceb-3123b2bb15d4.png)

Sending side:
![image](https://user-images.githubusercontent.com/121945/168469065-0dcf8516-0da4-4724-b3c8-c07953867290.png)

Reply:
For CLI I think the only thing to do would be to add a `--request` flag to `wormhole receive` that would generate this code, and then wait for connection and save the file

I don't know how the CLI would want to support the URI scheme on the parsing side. It might allow URI entry for the code, which would prompt to enter a file name if the code direction would be reversed. This would not be the best workflow, but I also don't see any reason why someone would want to do this with the CLI apart from doing it by accident.

Reply:
Maybe haven't thought this _completely_ through, but it could also work if the CLI did `send` along with a "request" URI -- that is, semantically you're still "sending" from the CLI side.

The Python CLI already has some interactive modes, so it _could_ prompt for a filename (although I agree this is a little weird for CLIs and doing e.g. tab-completion would probably be "expected" but also non-trivial).

So basically something like `wormhole send --uri <valid request URI> /path/to/local/file/to/send`

Reply:
I think for the Rust implementation, I'll go with something like `wormhole receive-request` which gives you a code. Then, the sender will hat to do `wormhole send --code`, which is the same thing as using custom code which is supported already. It's important to me that the URIs behave identically to codes, so at any point in time one may choose to use the one or the other.

That being said, let's pain that shed: For transit, we use "leader" and "follower" roles instead of "sender" and "receiver" to make more clear it is protocol agnostic. Also "type" may be a bit confusing (see @meejah's comment below), so how about "role" or "direction"?

Reply:
`role=leader` means the URI parsing application would have to ... do something? Send a file? Decide what it wants to do instead of executing something predetermined? Not sure yet what it would mean in practice.

What about a potential new mode that would allow for file transfer in both directions? Who would be who in that case? I don't propose we add that here right away, just wondering how it could look like.